core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -761,13 +809,16 @@ class GroupCoordinator(val brokerId: Int,
   private def addMemberAndRebalance(rebalanceTimeoutMs: Int,
                                     sessionTimeoutMs: Int,
                                     memberId: String,
+                                    groupInstanceId: Option[String],
                                     clientId: String,
                                     clientHost: String,
                                     protocolType: String,
                                     protocols: List[(String, Array[Byte])],
                                     group: GroupMetadata,
                                     callback: JoinCallback) {
-    val member = new MemberMetadata(memberId, group.groupId, clientId, clientHost, rebalanceTimeoutMs,
+    val member = new MemberMetadata(memberId, group.groupId,
+      groupInstanceId.getOrElse(JoinGroupRequest.EMPTY_GROUP_INSTANCE_ID),
```

Comment:
Why not let `groupInstanceId` be represented as an `Option` inside `MemberMetadata` as well?

Reply:
I tried one time and it failed due to serialization issue. Let me try one more time.

Reply:
To make the full e2e consistency, we should consider supporting Optional[String] type for part of auto-mated protocol in `JoinGroupRequest.json` and other protocol classes. Otherwise, we still need to have `EMPTY_INSTANCE_ID` as a special type to handle null case in the serde of request. @hachikuji

Reply:
The automated protocol supports `nullable string` (it will be serialized and stored as `0xffff` over the wire), and hence could we encode null for this instance id, and then:

1) on client side we can have this parameter nullable, and upon constructing join-group request it will be auto-serialized.
2) on broker side we can have this parameter as `Option[String]`, upon deserializing if the returned value is null construct the field as `None`.

Reply:
@guozhangwang Thanks for the info!

Reply:
Sorry, if this wasn't clear, but here is what you need to add to the schema definition:
```
    { "name": "GroupInstanceId", "type": "string", "versions": "5+", "nullableVersions": "5+",
      "about": "The unique identifier of the consumer instance provided by end user." },
```

Reply:
Yea, I just figured it out, thank you!! @hachikuji